contracts/router/modules/pool/uniswap/UniswapV3Module.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {IndexedToken, IPoolModule} from "../../../interfaces/IPoolModule.sol";
+import {IUniswapV3Pair} from "../../../interfaces/uniswap/IUniswapV3Pair.sol";
+import {ExactInputSingleParams, IUniswapV3Router} from "../../../interfaces/uniswap/IUniswapV3Router.sol";
+// prettier-ignore
+import {
+    QuoteExactInputSingleParams, IUniswapV3StaticQuoter
+} from "../../../interfaces/uniswap/IUniswapV3StaticQuoter.sol";
+import {OnlyDelegateCall} from "../OnlyDelegateCall.sol";
+
+import {SafeERC20, IERC20} from "@openzeppelin/contracts-4.5.0/token/ERC20/utils/SafeERC20.sol";
+
+contract UniswapV3Module is OnlyDelegateCall, IPoolModule {
+    using SafeERC20 for IERC20;
+
+    /// These need to be immutable in order to be accessed via delegatecall
+    IUniswapV3Router public immutable uniswapV3Router;
+    IUniswapV3StaticQuoter public immutable uniswapV3StaticQuoter;
+
+    constructor(address uniswapV3Router_, address uniswapV3StaticQuoter_) {
+        uniswapV3Router = IUniswapV3Router(uniswapV3Router_);
+        uniswapV3StaticQuoter = IUniswapV3StaticQuoter(uniswapV3StaticQuoter_);
+    }
+
+    /// @inheritdoc IPoolModule
+    function poolSwap(
+        address pool,
+        IndexedToken memory tokenFrom,
+        IndexedToken memory tokenTo,
+        uint256 amountIn
+    ) external returns (uint256 amountOut) {
+        // This function should be only called via delegatecall
+        assertDelegateCall();
```

Comment:
What's the logic for only allowing `poolSwap` to be called via delegatecall?

Reply:
This is done to enforce the correct usage of the Module: have custody over tokens, then delegate call to Module.

As opposed to "transfer tokens to Module, then external call to Module" which would also work, if no delegate call check is done.